test/e2e/arktos/multi_tenancy/cmd/framework/testcase.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2020 Authors of Arktos.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+type TestCase struct {
+	Command string `yaml:"Command,omitempty"`
+	TimeOut *int   `yaml:"TimeOut,omitempty"`
+
+	RunBefore string `yaml:"RunBefore,omitempty"`
+
+	RetryCount    *int `yaml:"RetryCount,omitempty"`
+	RetryInterval *int `yaml:"RetryInterval,omitempty"`
+
+	TerminateAfter *int `yaml:"TerminateAfter,omitempty"`
+
+	ShouldFail          bool     `yaml:"ShouldFail,omitempty"`
+	OutputShouldBe      string   `yaml:"OutputShouldBe,omitempty"`
+	OutputShouldContain []string `yaml:"OutputShouldContain,omitempty"`
+
+	OutputShouldNotContain []string `yaml:"OutputShouldNotContain,omitempty"`
+}
+
+func (t *TestCase) Validate(tc *TestConfig) *ErrorList {
+	errList := NewErrorList()
+
+	errList.Add(ErrorIfEmpty("Command", t.Command))
+	errList.Add(ErrorIfNotWithin("RetryCount", t.RetryCount, 0, tc.MaxRetryCount))
+	errList.Add(ErrorIfNotWithin("RetryInterval", t.RetryInterval, 0, tc.MaxRetryInterval))
+	errList.Add(ErrorIfNotWithin("TimeOut", t.TimeOut, 0, tc.MaxTimeOut))
+
+	return errList
+}
+
+func (t *TestCase) Complete(tc *TestConfig) {
+	SetDefaultIfNil(&t.RetryCount, &tc.DefaultRetryCount)
+	SetDefaultIfNil(&t.RetryInterval, &tc.DefaultRetryInterval)
+	SetDefaultIfNil(&t.TimeOut, &tc.DefaultTimeOut)
+}
+
+func (t *TestCase) Run(tc *TestConfig) *ErrorList {
+	errList := NewErrorList()
+	var testErrList *ErrorList
+	var exitCode int
+	var output string
+	var err error
+
+	t.Complete(tc)
+	if t.RunBefore != "" {
+		LogWarning("Run pre-test configuration command %q", t.RunBefore)
+		exitCode, output, err = ExecCommandLine(t.RunBefore, 0)
+		if exitCode != 0 {
+			// it is by design that pre-test configuration command is allowed to fail
+			LogWarning("Pre-test command %q returns with error: %v, exit code: %v, command output: %q. Moving on ...", t.RunBefore, exitCode, output)
+		}
+	}
+
+	retries := *t.RetryCount
+	LogNormal("Testing %q...", t.Command)
+	for {
+		exitCode, output, err = ExecCommandLine(t.Command, *t.TimeOut)
+		if tc.Verbose {
+			LogNormal("\nexit code: %v, output:\n%v", exitCode, output)
+		}
+
+		if err == nil {
+			testErrList = t.CheckTestResult(exitCode, output)
+			if testErrList.IsEmpty() {
+				LogSuccess(" PASSED\n")
+				break
+			}
+		}
+
+		if retries <= 0 {
```

Comment:
not a big fan of 3-nested loops. we want as few nesting as possible. easier to read and test. here's my suggestion:

```golang
if retries > 0 {
		LogWarning("\nFailed, retrying after %v seconds ...", *t.RetryInterval)
		retries--
		time.Sleep(time.Duration(*t.RetryInterval) * time.Second)
		continue;
}

if err != nil {
		errList.Add(err)
		LogError("\nFAILED: %v\n", err)
                 break;
}

errList.Concat(testErrList)
LogError("\nFAILED: %v \n", testErrList.String())
if !tc.Verbose {
		LogNormal("exit code: %v, output:\n%v", exitCode, output)
}
break
```

Reply:
refactored.